### ISSUE TITLE
Table of Contents

### ISSUE THREAD

**Issue by Ciantic** [RelaxedJS/ReLaXed#8]
Is it possible to generate page numbers to toc, and having toc rows clickable? Also clickable references section is required.

That was so hard that I myself resorted to LaTeX instead of Chrome few months ago (didn't know about ReLaXed back then though).

Basically to support those one needs to do rendering in multiple passes, which is the reason LaTeX is so damn slow sometimes. It renders the page, get's the page numbers, make changes, re-renders the page and ensures the page numbers didn't change etc.

**Comment by Zulko** [comment]
Yes for "having clickable toc rows", but as for page number id say it is almost certainly impossible. Some PDF viewers will have a sidebar with sections and page number, but that won't be of any use if you intend to print your document.

**Comment by Ciantic** [comment]
I have some crazy idea:

Have all numbers e.g. 1-300 overlayed top of each other, and post-processing style remove the unneeded once from the PDF document by looking at the underlying link to section.

Not expecting you to implement this, but I leave this here as I might myself investigate this.

**Comment by benperiton** [comment]
So I had a bit of a play with this earlier, and my initial thinking was to use a Pug plugin and work on the AST that is generated, but it seems that the filters generate Text nodes once they are run, so if there are headings in a ```:markdown-it section```, then they don't become part of it, which is a shame because that would have been a nice way to potentially do citations etc as well.

My current thinking is that it would have to done once the HTML is generated.

@Zulko Did you have any ideas currently on this? I guess an automatic TOC would need to take into account all ways of generating a <h?> tag, rather than just pug?

**Comment by Zulko** [comment]
I believe the safest way would be some javascript exploring the DOM when all the rest of the page has finished rendering (probably a jquery plugin). Still, this would give you a list of links to sections, but not the page number.

**Comment by Drew-S** [comment]
I created a working Table of Contents system: [ToC](https://github.com/Drew-S/ReLaXed/tree/TableOfContents). It is still a bit of a work in progress, but it works fairly well, there are some slight improvements to the workings of the bibliography as well, though no new features there.

**Comment by Zulko** [comment]
Note That there is already a TOC in ReLaXed documents, just not printable (some PDF viewers will show it in the sidebar). For an in-document TOC, there are many many cave-ats related to page numbers, including the fact that the TOC itself may shift the page by one or two, the fact that the page numbers have to be evaluated in "print" layout mode, The fact that they will depend on the page height, but also accounting for left right top bottom margins which are set at PDF writing time. And a broken TOC will bring us more bug fix requests than no TOC at all. So I am a bit torn by this issue right now.

**Comment by Drew-S** [comment]
I am not seeing where the ToC is generated, I assume puppeteer generates it then?

So far my ToC is not complete, but it is printable, it generates the text where the `+ToC` is called then goes through and obtains the page numbers, so the ToC shifts everything down before finishing the page numbers. It also currently takes into account the page format, but it does not take into account the page margins, so its page numbering is not correct yet.

**Comment by Drew-S** [comment]
I did some more improvements to the ToC, still need to do plenty more testing to ensure its working fully. The ToC generation takes into account: the ToC shifting content, the top, right, left, and bottom margins, the height and width of the pages.

It does this by, when generating the ToC, force the body width to be the print width (8.5in) minus the left and right margins, it then increments a page number every time an element to mark is seen above an integer of the height minus the top and bottom margins.

**Comment by upstroke** [comment]
Hi @Drew-S 
So is there an update on the TOC and when is ported to a plugin we can use?
Would be very useful to have. 
THX

**Comment by Zulko** [comment]
I have kind of called for a break on this feature as it doesn't play well with Chromium (there are cases that make it difficult to predict which page an element will en up on). See the end of this thread for the latest discussion:

https://github.com/RelaxedJS/ReLaXed/issues/88

I believe there is a possible solution using a processing of the PDF itself, but I didnt come around to test it yet.

**Comment by Drew-S** [comment]
Unfortunately, I have been busy the last week and a half so have not done any work on any of this. As @Zulko has stated, it is placed on pause, as there is formatting issues that cannot be worked around. I spent some time looking through puppeteer issues. [#1778](https://github.com/GoogleChrome/puppeteer/issues/1778) is a request for ToC features in Puppeteer itself. I have opened up an issue on puppeteer for adding a hook into `page.pdf()` to be able to manipulate the DOM after it has been formatted for print [#2773](https://github.com/GoogleChrome/puppeteer/issues/2773). If we get something like that then we could do most everything we need, ToC, Figure list, custom footers, etc.

**Comment by Ciantic** [comment]
Problem is this: Pupeteer team is not themselves trying to solve these problems. 

This means that when you open a issue there, they can't understand what hurdles you have hit trying to solve it. Clearly @Drew-S gave a good try on this, and found a deficiency in Puppeteer, but they will dismiss these API changes for as long as they themselves try to do same.

E.g. the issue they asked people to make to Chromium ([840455](https://bugs.chromium.org/p/chromium/issues/detail?id=840455)) is a joke. Will they make a one fits for all TOC? It makes no sense, none wants fully baked TOC in API, we need extensible API such as modifying the DOM.

**Comment by Gee19** [comment]
@Drew-S is there a way to use the mixins you built in an HTML doc? I'm experimenting with your fork on branch `ToCplugin` and I've created a `config.yml` in my document root folder that contains:

`
plugins:
  - tableofcontents
`

Any help would be greatly appreciated :)

#109

**Comment by rfizzle** [comment]
Is it possible to get this plugin without requiring the page numbers?

**Comment by DanielRuf** [comment]
> Is it possible to get this plugin without requiring the page numbers?

Can you explain further what you need?

**Comment by rfizzle** [comment]
> > Is it possible to get this plugin without requiring the page numbers?
> 
> Can you explain further what you need?

I was able to fork the plugin Drew-S created. I just needed a digital ToC that auto generated inline and had links to headers for my PDFs. Even without page numbers, it's better than nothing.

<img width="983" alt="image" src="https://user-images.githubusercontent.com/2577744/63398749-bc4d5080-c3f8-11e9-8625-4bba7dc7cb7c.png">

**Comment by DanielRuf** [comment]
Looks cool. Do you want to open a PR to add this?
